### PR TITLE
[.yeb] easyconfig files in YAML syntax working with dependencies

### PR DESCRIPTION
**Pull request by Caylo**

### CODE REVIEW

**Comment by boegel** [comment]
Jenkins: ok to test

**Comment by boegel** [comment]
@wpoely86, @rjeschmi: please review?

File: easybuild/framework/easyconfig/format/yeb.py [hunk 1]
```diff
@@ -127,3 +127,13 @@ def is_yeb_format(filename, rawcontent):
             if line.startswith('name: '):
                 isyeb = True
     return isyeb
+
+# join functionality:
+# from http://stackoverflow.com/questions/5484016/how-can-i-do-string-concatenation-or-string-replacement-in-yaml/23212524#23212524
+# define custom tag handler
+def join(loader, node):
```
Comment (wpoely86) [inline easybuild/framework/easyconfig/format/yeb.py]: always document the arguments

File: easybuild/framework/easyconfig/format/yeb.py [hunk 2]
```diff
@@ -127,3 +127,13 @@ def is_yeb_format(filename, rawcontent):
             if line.startswith('name: '):
                 isyeb = True
     return isyeb
+
+# join functionality:
```
Comment (wpoely86) [inline easybuild/framework/easyconfig/format/yeb.py]: maybe put this in the docstring? @boegel do we have a policy for this?

Comment (boegel) [inline easybuild/framework/easyconfig/format/yeb.py]: description of the function should indeed go into the docstring, references to details w.r.t. the implementation should be comments

File: easybuild/framework/easyconfig/format/yeb.py (hunk 1 above)
Comment (boegel) [inline easybuild/framework/easyconfig/format/yeb.py]: `join` is maybe a bit too generic of a name for this function, maybe use `yaml_join`?

File: test/framework/easyconfigs/yeb/goolf-1.4.10.yeb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    - &comp_version 4.7.2
+    - &comp [*comp_name, *comp_version]
+
+    - &blaslib OpenBLAS
+    - &blasver 0.2.6
+    - &blas !join [*blaslib, -, *blasver]
+    - &blas_suff -LAPACK-3.4.2
+
+    - &comp_mpi_tc_name gompi
+    - &comp_mpi_tc_ver 1.4.10 #FIXME version template
```
Comment (boegel) [inline test/framework/easyconfigs/yeb/goolf-1.4.10.yeb]: this should be fixed before we merge this, i.e. use `*version` here?

File: test/framework/easyconfigs/yeb/goolf-1.4.10.yeb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+version: 1.4.10
+
+homepage: (none)
+description: |
+    GNU Compiler Collection (GCC) based compiler toolchain, including
+     OpenMPI for MPI support, OpenBLAS (BLAS and LAPACK support), FFTW and ScaLAPACK.
+
+toolchain: {name: dummy, version: dummy}
+
+vars:
```
Comment (boegel) [inline test/framework/easyconfigs/yeb/goolf-1.4.10.yeb]: defining these variables should be done on top, before actually defining easyconfig parameters?

maybe we should use something like `_internal_variables_` as 'key' here (just as a style thing), to make it clear?

File: test/framework/easyconfigs/yeb/intel-2015b.yeb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+toolchain: {'name': 'dummy', 'version': 'dummy'}
+
+vars:
+    - &compver 2015.3.187
+    - &gccsuff -GNU-4.9.3-2.25
+
+dependencies: [
+    [icc, *compver, *gccsuff],
+    [ifort, *compver, *gccsuff],
+    [impi, 5.0.3.048, '', [iccifort, !join [*compver, *gccsuff]]],
```
Comment (boegel) [inline test/framework/easyconfigs/yeb/intel-2015b.yeb]: would `!join *compver, *gccsuff` also work? (less square brackets to worry about)

File: test/framework/yeb.py [hunk 6]
```diff
@@ -61,6 +61,8 @@ def test_parse_yeb(self):
         test_files = {
             'bzip2.yeb': 'bzip2-1.0.6-GCC-4.9.2.eb',
             'gzip.yeb': 'gzip-1.6-gcc-4.9.2.eb',
+            'intel-2015b.yeb': 'intel-2015b.eb',
```
Comment (boegel) [inline test/framework/yeb.py]: maybe it's better to create the `.yeb` equivalent of `ictce-4.1.13.eb`, which is already in the test easyconfigs

add `intel-2015b.eb` requires also adding easyconfigs for each of the components (`icc`, `ifort`, etc.) to make the tests happy

**Comment by boegel** [comment]
@Caylo: remove the `intel-2015b.eb` test easyconfigs, since it's not used and breaking the tests ;)

File: test/framework/easyconfigs/yeb/goolf-1.4.10.yeb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    - &comp_name GCC
+    - &comp_version 4.7.2
+    - &comp [*comp_name, *comp_version]
+
+    - &blaslib OpenBLAS
+    - &blasver 0.2.6
+    - &blas !join [*blaslib, -, *blasver]
+    - &blas_suff -LAPACK-3.4.2
+
+    - &comp_mpi_tc_name gompi
```
Comment (boegel) [inline test/framework/easyconfigs/yeb/goolf-1.4.10.yeb]: not used anywhere else, so just use `gompi` in the line below and drop this one?

File: test/framework/easyconfigs/yeb/goolf-1.4.10.yeb [hunk 8]
```diff
@@ -0,0 +1,39 @@
+_internal_variables_:
+    - &version 1.4.10
+    - &comp_name GCC
```
Comment (boegel) [inline test/framework/easyconfigs/yeb/goolf-1.4.10.yeb]: insert empty line above, the `version` anchor/alias doesn't relate to `GCC`

File: test/framework/easyconfigs/yeb/goolf-1.4.10.yeb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+easyblock: Toolchain
+
+name: goolf
+version: *version
+
+homepage: (none)
+description: |
+    GNU Compiler Collection (GCC) based compiler toolchain, including
+     OpenMPI for MPI support, OpenBLAS (BLAS and LAPACK support), FFTW and ScaLAPACK.
```
Comment (boegel) [inline test/framework/easyconfigs/yeb/goolf-1.4.10.yeb]: align with line above? (lose on space)

File: test/framework/easyconfigs/yeb/ictce-4.1.13.yeb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+name: ictce
+version: 4.1.13
+
+homepage: http://software.intel.com/en-us/intel-cluster-toolkit-compiler/
+description: Intel Cluster Toolkit Compiler Edition provides Intel C/C++ and Fortran compilers, Intel MPI & Intel MKL.
+
+toolchain: {name: dummy, version: dummy}
+
+
```
Comment (boegel) [inline test/framework/easyconfigs/yeb/ictce-4.1.13.yeb]: drop one empty line above

File: test/framework/yeb.py [hunk 11]
```diff
@@ -94,6 +96,18 @@ def test_is_yeb_format(self):
         self.assertFalse(is_yeb_format(test_eb, None))
         self.assertFalse(is_yeb_format(None, raw_eb))
 
+
+    def test_join(self):
+        """ Test yaml_join function """
+
+        stream = "hw: !join [hello, world]"
```
Comment (boegel) [inline test/framework/yeb.py]: also use aliases/variables in the test for `!join`?

like you show in the example in the docs

**Comment by boegel** [comment]
Thanks @Caylo!

**Pull request merged**